sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp

```diff
@@ -77,68 +101,159 @@ constexpr int get_layout_id<
                                             matrix::matrix_layout::col_major>> {
   void load(sycl::ext::oneapi::experimental::matrix::joint_matrix<
-                double, sycl::ext::oneapi::experimental::matrix::matrix_use::b,
-                4, 8, Layout> &res,
-            multi_ptr<double, Space> src, size_t stride) {
+                T, Use, NumRows, NumCols, Layout> &res,
+            multi_ptr<T, Space> src, size_t stride) {
 #ifdef __NVPTX__
 #ifdef __SYCL_DEVICE_ONLY__
-    __dmma_m8n8k4_ld_b(res.data, src.get(), stride, get_layout_id<Layout>());
-#endif
-#endif
-  }
-};
 
-template <sycl::ext::oneapi::experimental::matrix::matrix_layout Layout,
-          access::address_space Space>
-struct joint_matrix_load_impl<
-    double, sycl::ext::oneapi::experimental::matrix::matrix_use::accumulator, 8,
-    8, Layout, Space,
-    typename std::enable_if_t<Layout == sycl::ext::oneapi::experimental::
-                                            matrix::matrix_layout::row_major ||
-                              Layout == sycl::ext::oneapi::experimental::
-                                            matrix::matrix_layout::col_major>> {
-  void
-  load(sycl::ext::oneapi::experimental::matrix::joint_matrix<
-           double,
-           sycl::ext::oneapi::experimental::matrix::matrix_use::accumulator, 8,
-           8, Layout> &res,
-       multi_ptr<double, Space> src, size_t stride) {
-
-#ifdef __NVPTX__
-#ifdef __SYCL_DEVICE_ONLY__
-    __dmma_m8n8k4_ld_c(res.data, src.get(), stride, get_layout_id<Layout>());
+    if constexpr (std::is_same<T, uint16_t>::value) {
+      int32_t *tileptr = reinterpret_cast<int32_t *>(src.get());
+      if constexpr (NumRows == 16 && NumCols == 16) {
+        if constexpr (Use ==
+                      sycl::ext::oneapi::experimental::matrix::matrix_use::a) {
+          __mma_bf16_m16n16k16_ld_a(res.data, tileptr, stride,
+                                    get_layout_id<Layout>());
+        } else if constexpr (Use == sycl::ext::oneapi::experimental::matrix::
+                                        matrix_use::b) {
+          __mma_bf16_m16n16k16_ld_b(res.data, tileptr, stride,
+                                    get_layout_id<Layout>());
+        }
+      } else if constexpr (NumRows == 8 && NumCols == 16) {
+        __mma_bf16_m8n32k16_ld_a(res.data, tileptr, stride,
+                                 get_layout_id<Layout>());
+      } else if constexpr (NumRows == 16 && NumCols == 32) {
+        __mma_bf16_m8n32k16_ld_b(res.data, tileptr, stride,
+                                 get_layout_id<Layout>());
+      } else if constexpr (NumRows == 32 && NumCols == 16) {
+        __mma_bf16_m32n8k16_ld_a(res.data, tileptr, stride,
+                                 get_layout_id<Layout>());
+      } else if constexpr (NumRows == 16 && NumCols == 8) {
+        __mma_bf16_m32n8k16_ld_b(res.data, tileptr, stride,
+                                 get_layout_id<Layout>());
+      }
```

Comment:
Thanks

Reply:
In joint_matrix_load_impl, joint_matrix_store_impl, and joint_matrix_mad_impl the user must provide joint_matrix(s) that is constructed with a sycl::sub_group template parameter.  So long as I implement cases for all eligible joint_matrix and I ensure that there is a 1:1 mapping between all eligible joint_matrix and the constexpr statements/builtin calls within these functions, there is no need to have any error cases.

Errors are only possible when constructing joint_matrix.  I looked at the option of providing a default constructor for joint_matrix using the sycl::sub_group template parameter and then using an assert statement as appropriate for unsupported instances of joint_matrix (similar to the intel matrix implementation's use of assert).  However the error message returned using this method was completely ambiguous.  Therefore I reverted to the existing case where if a user tries to construct an unsupported joint_matrix then they get the error that the constructor doesn't exist.
It could be better to use an exception system that can provide the user with a simpler and more concise error message such as "invalid joint_matrix"; however this is not achieved using assert, and I think that we should decide on an error message system that is consistent across backend implementations first.

Reply:
What do you mean by ambiguous?
 What will something like static_assert((NumRows != X &&
                   NumCols != Y),
                  "Tensorcores matrix implementation does not support these sizes; refer to ... for supported sizes");
  }

Reply:
> What do you mean by ambiguous? What will something like static_assert((NumRows != X && NumCols != Y), "Tensorcores matrix implementation does not support these sizes; refer to ... for supported sizes"); }

If I use a runtime assert like [in matrix-jit.hpp](https://github.com/intel/llvm/blob/29047bc97fae34bdbc0a4fb936f4429f78c6c8f1/sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp#L84) then I get a general cuda assert exception with lots of lines of thread assert failures that give zero information about the specific joint_matrix failure.  An example line is:
```
:43: block: [0,1,0], thread: [31,0,0] Assertion `` failed.
```

I tried using the compile time static_assert that you suggested (thanks for the suggestion, I had not used static_assert before) but a problem with this is that if I use the correct assert boolean the error message will be very very long: I'd need to check each of the possible cases as used by the macro "__SYCL_JOINT_MATRIX_OVERLOAD".  I could instead have a default constructor for joint_matrix to cover cases not enumerated by the __SYCL_JOINT_MATRIX_OVERLOAD macro, but I would have to use one of the template parameters in the assert to ensure that the correct constructor is called by the compile time assert, and the assert would not relate to the error which could be confusing for the user, e.g. if I choose to use the Group template parameter I get an error like:

```
static_assert((std::is_same<Group, sycl::sub_group>::value && false), "Tensor Cores joint_matrix implementation does not support these sizes for the requested data type");
```

I think that this is worse than the error messages that I get from the existing implementation where the constructor is only implemented for supported cases, which is correct and clear:
```
joint_matrix_tensorcore.cpp:156:83: error: implicit instantiation of undefined template 'sycl::ext::oneapi::experimental::matrix::joint_matrix<sycl::detail::half_impl::half, sycl::ext::oneapi::experimental::matrix::matrix_use::a, 14, 16, sycl::ext::oneapi::experimental::matrix::matrix_layout::row_major>'
joint_matrix<T1, matrix_use::a, 14, K, matrix_layout::row_major, sycl::sub_group> sub_ze;
                                                                                  ^
joint_matrix_tensorcore.cpp:193:3: note: in instantiation of function template specialization 'test<sycl::detail::half_impl::half, float, 2UL, 4UL, 3UL, 16UL, 16UL, 16UL>' requested here
  test<half, float, SUB_TILES_M, SUB_TILES_K, SUB_TILES_N, 16, 16, 16>();
  ^
/home/jackkirk/jkLLVM/llvm/build/bin/../include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp:19:8: note: template is declared here
struct joint_matrix;
       ^
1 error generated.
```

I'm open to suggestions if you think there is a better option? Thanks

Reply:
With the static_cast, you can improve the message to tell the user how to fix it. But with the current approach, it will be hard to debug as the user does not know whether they made a typo or the sizes are not supported but this is a problem of C++ metaprogramming, it is hard to debug :) 
On the other hand, we state that the query interface is the tool to use to know what is supported and what is not, so errors are not mandatory here. It is just that since this is AOT, we can take advantage of some of compile-time errors that we cannot currently do with matrix-jit since it uses JIT compilation.

Reply:
Yes I was thinking that the query interface was more the place to find out suitable types/shapes.  I agree that it would be best to have a short message informing the user that the chosen type/shape is not supported, but with the current implementation I do not know you to do this using static_assert (without making the assert unrelated to the actual error), since it would output a very large boolean assert that effectively spits out in one go all of the information that the query interface would provide.  I have no strong feelings on this however and would be happy to switch to the best possible static_assert for the current implementation if you want me to.

I think that once we are at the point of merging the dpas/Nvidia backend implementations we should reconsider the error messaging and ensure that both messaging systems (and messages) are consistent and effective.  
I personally think that at the moment it is the correct tactic to keep the backends in separate headers, but to each keep the other backend implementation in mind, and conform to the same proposal.  So far I think that both implementations could immediately be merged together as they are (e.g. #ifdef NVPTX sends joint_matrix_load to joint_matrix_load_impl, and DPAS is sent to the intel implementation functions).  So as long as both are constrained to conform closely to the extension proposal, keeping them separate allows some small degree of freedom and experimentation in both backends, in case of unforeseen issues that we encounter from time to time, to ensure that all features of both backends can be supported to the full extent desired before a final universal design is decided upon.